lib/sequencer/src/execution/block_executor.rs

```diff
@@ -177,8 +177,20 @@ pub async fn execute_block<R: ReadStateHistory + WriteState>(
                                                 tracing::warn!(tx_hash = %tx.hash(), block = ctx.block_number, ?e, "invalid tx → skipped");
                                             },
                                             TxRejectionMethod::SealBlock(reason) => {
-                                                tracing::debug!(tx_hash = %tx.hash(), block = ctx.block_number, ?e, ?reason, "sealing block by criterion");
-                                                break reason;
+                                                // For ProduceBlock (SealPolicy::Decide) or Replay (SealPolicy::UntilExhausted { allowed_to_finish_early: false }), don't seal if no transactions have been executed yet
+                                                if matches!(command.seal_policy, SealPolicy::Decide(..) | SealPolicy::UntilExhausted { allowed_to_finish_early: false }) && executed_txs.is_empty() {
```

Comment:
I think you shouldn't include SealPolicy::UntilExhausted { allowed_to_finish_early: false }) here, just let it through, it will return Err here https://github.com/matter-labs/zksync-os-server/pull/653/files#diff-aedb1011e6be536b7f6ff66eef94cb732bb4fe84a615736edbb9e37ca88cc83fR248